File: packages/shared-ts/src/logger.ts
/**
 * Structured logger that writes to:
 *   1. stdout (JSON lines — Cloud Run log viewer works unchanged)
 *   2. Upstash Redis Stream `logs:{stage}` (for the `pnpm logs` CLI)
 *
 * XADD is fire-and-forget. If Redis is unreachable, the stdout side still
 * records the event; logging must never crash the caller.
 */

import type { LogEntry, LogLevel } from "@annemusic/contracts";
import { redis } from "./redis.js";

const MAX_STREAM_LEN = 10_000;

type Fields = Record<string, unknown>;

function errObj(err: unknown): LogEntry["err"] {
  if (!err) return undefined;
  if (err instanceof Error) return { name: err.name, message: err.message, stack: err.stack };
  return { name: "NonError", message: String(err) };
}

function flatten(entry: LogEntry): Record<string, string> {
  // Upstash XADD takes a shallow string→string map; serialize nested fields.
  return {
    ts: String(entry.ts),
    stage: entry.stage,
    job_id: entry.job_id ?? "",
    level: entry.level,
    msg: entry.msg,
    data: entry.data ? JSON.stringify(entry.data) : "",
    err: entry.err ? JSON.stringify(entry.err) : "",
  };
}

async function publish(entry: LogEntry) {
  const key = `logs:${entry.stage}`;
  try {
    await redis().xadd(key, "*", flatten(entry), {
      trim: { type: "MAXLEN", threshold: MAX_STREAM_LEN, comparison: "~" },
    });
  } catch (e) {
    // Structured warning so the failure is itself searchable.
    console.error(JSON.stringify({
      ts: Date.now(),
      stage: entry.stage,
      level: "warn",
      msg: "log stream publish failed",
      err: errObj(e),
    }));
  }
}

export type Logger = {
  debug: (job_id: string | undefined, msg: string, data?: Fields) => void;
  info: (job_id: string | undefined, msg: string, data?: Fields) => void;
  warn: (job_id: string | undefined, msg: string, data?: Fields) => void;
  error: (job_id: string | undefined, msg: string, err: unknown, data?: Fields) => void;
};

export function createLogger(stage: string): Logger {
  const emit = (level: LogLevel, job_id: string | undefined, msg: string, data?: Fields, err?: unknown) => {
    const entry: LogEntry = {
      ts: Date.now(),
      stage,
      job_id,
      level,
      msg,
      data,
      err: err ? errObj(err) : undefined,
    };
    // stdout first — never blocks on the network.
    process.stdout.write(JSON.stringify(entry) + "\n");
    // Stream in the background.
    void publish(entry);
  };
  return {
    debug: (job_id, msg, data) => emit("debug", job_id, msg, data),
    info: (job_id, msg, data) => emit("info", job_id, msg, data),
    warn: (job_id, msg, data) => emit("warn", job_id, msg, data),
    error: (job_id, msg, err, data) => emit("error", job_id, msg, data, err),
  };
}

/** Await any in-flight XADDs. Call before process exit in short-lived scripts. */
export async function flushLogs(): Promise<void> {
  // XADDs in this module are all fire-and-forget Promises. We rely on the
  // pnpm/node runtime to await them; in short-lived scripts we also sleep
  // briefly to give the HTTP request time to land.
  await new Promise((r) => setTimeout(r, 200));
}

File: packages/shared-py/shared/logger.py
"""Mirror of shared-ts logger: stdout JSON lines + Upstash Redis Stream.

Same contract (LogEntry) as the TS side so `pnpm logs --stage X --follow` sees
entries from Python stages identically.
"""

from __future__ import annotations

import json
import sys
import time
import traceback
from dataclasses import dataclass
from typing import Any, Callable

from .redis_client import redis

MAX_STREAM_LEN = 10_000
LEVELS = ("debug", "info", "warn", "error")


def _err_obj(err: BaseException | None) -> dict[str, str] | None:
    if err is None:
        return None
    return {
        "name": type(err).__name__,
        "message": str(err),
        "stack": "".join(traceback.format_exception(type(err), err, err.__traceback__)),
    }


def _flatten(entry: dict[str, Any]) -> dict[str, str]:
    return {
        "ts": str(entry["ts"]),
        "stage": entry["stage"],
        "job_id": entry.get("job_id") or "",
        "level": entry["level"],
        "msg": entry["msg"],
        "data": json.dumps(entry["data"], separators=(",", ":")) if entry.get("data") else "",
        "err": json.dumps(entry["err"], separators=(",", ":")) if entry.get("err") else "",
    }


def _publish(entry: dict[str, Any]) -> None:
    try:
        redis().xadd(
            f"logs:{entry['stage']}",
            {"values": _flatten(entry)},  # upstash_redis Python: hset-style values= kwarg-free API
            # NOTE: upstash_redis python xadd signature is (name, id, values, ...).
            # We use id="*" and values=dict; MAXLEN trim via approximate.
        )
    except TypeError:
        # Fallback if signature differs in newer versions — positional call.
        try:
            redis().xadd(f"logs:{entry['stage']}", "*", _flatten(entry))
        except Exception as e:  # pragma: no cover — never crash the caller
            sys.stderr.write(json.dumps({
                "ts": int(time.time() * 1000),
                "stage": entry["stage"],
                "level": "warn",
                "msg": "log stream publish failed",
                "err": _err_obj(e),
            }) + "\n")
    except Exception as e:  # pragma: no cover
        sys.stderr.write(json.dumps({
            "ts": int(time.time() * 1000),
            "stage": entry["stage"],
            "level": "warn",
            "msg": "log stream publish failed",
            "err": _err_obj(e),
        }) + "\n")


@dataclass
class Logger:
    stage: str

    def _emit(
        self,
        level: str,
        job_id: str | None,
        msg: str,
        data: dict[str, Any] | None = None,
        err: BaseException | None = None,
    ) -> None:
        entry = {
            "ts": int(time.time() * 1000),
            "stage": self.stage,
            "job_id": job_id,
            "level": level,
            "msg": msg,
            "data": data,
            "err": _err_obj(err),
        }
        sys.stdout.write(json.dumps(entry) + "\n")
        sys.stdout.flush()
        _publish(entry)

    def debug(self, job_id: str | None, msg: str, data: dict[str, Any] | None = None) -> None:
        self._emit("debug", job_id, msg, data)

    def info(self, job_id: str | None, msg: str, data: dict[str, Any] | None = None) -> None:
        self._emit("info", job_id, msg, data)

    def warn(self, job_id: str | None, msg: str, data: dict[str, Any] | None = None) -> None:
        self._emit("warn", job_id, msg, data)

    def error(
        self,
        job_id: str | None,
        msg: str,
        err: BaseException,
        data: dict[str, Any] | None = None,
    ) -> None:
        self._emit("error", job_id, msg, data, err)


def create_logger(stage: str) -> Logger:
    return Logger(stage=stage)


def flush_logs() -> None:
    """Give in-flight xadds a moment to land. Call before exit in short scripts."""
    time.sleep(0.25)
